src/dir_status.ml

```diff
@@ -92,10 +92,15 @@ module DB = struct
       let project_root =
         File_tree.Dir.project ft_dir
         |> Dune_project.root
-        |> Path.source in
+      in
+      let build_dir_is_project_root =
+        match Path.Build.drop_build_context dir with
+        | None -> false (* assert false? *)
+        | Some p -> Path.Source.equal p project_root
```

Comment:
@aalekseyev this is the bit to pay attention to, I'm not entirely sure my diagnosis is correct but the old code looked quite suspicious to me.

Reply:
Yh we could indeed assert false here. The `None` case correspond to just `_build` which clearly contains nothing interesting from the point of view of `Dir_contents`.

Reply:
Alright, I got rid of the `None` case and turned it into a fatal error.